REFRAME.md

```diff
@@ -0,0 +1,291 @@
+    | "PutIPNSResponse" PutIPNSResponse
+    | "Error" Error
+}
+```
+
+Note: Each Request type has a corresponding Response type.
+Every message except the Error type should end in Request/Response.
+
+#### Error
+
+The Error message type should be used in Responses to indicate that an error has occurred.
+
+```ipldsch
+    type Error struct {
+        Method String
+    }
+```
+
+#### Identify
+
+A message for discovering which messages a server supports. May be used by applications to optimize which servers get sent which types of requests.
+
+```ipldsch
+    type IdentifyRequest struct {}
+
+    type IdentifyResponse struct {
+        Methods [String]
+    }
+```
+
+`IdentifyResponse` should return the set of supported methods aside from the "Identify" method.
+
+##### DAG-JSON Examples
+
+Request:
+```
+{"IdentifyRequest" : {}}
+```
+
+Response:
+```
+{"IdentifyResponse" : {
+    "Methods" : ["FindProviders", "GetIPNS"]
+}}
+```
+
+#### FindProviders
+
+A message for finding nodes that have an interest in a given key. Some common examples include finding which peers have advertised that they have a given CID, or which peers are interested in a given pubsub topic.
+
+```ipldsch
+    type FindProvidersRequest struct 
+        Key &Any
+    }
+```
+
+Note: While the Key is a CID it is highly recommended that server implementations treat these FindProvidersRequests as if they were for the multihash. This allows consumers that perform lookups by CID but with a different IPLD codec than one that is known to the server to still get back valid responses.
+
+```ipldsch
+    type FindProvidersResponse struct {
+        Providers [Provider]
+    }
+
+    type Provider struct {
+     Node Node
+     Proto optional [TransferProtocol]
+     }
+
+    # Node describes a node (identity, network address, authentication, whatever else applies)
+    # We expect different types of nodes, e.g. peer, miner, public IP, etc.
+    type Node union {
+     | Peer "peer"
+    } representation keyed
+
+    type Peer struct {
+        ID Bytes # The binary representation of a libp2p peerID https://github.com/libp2p/specs/blob/f433ad595224cf33d916c166d1738f11aadfa9f7/peer-ids/peer-ids.md
+        Multiaddresses optional [Bytes] # Each element in the list is the binary representation of a complete multiaddr without a peerID suffix
+                                        # https://github.com/multiformats/multiaddr/blob/d02c681171629912dd977f4f5d4b3b72dcac507f/README.md
+                                        # https://github.com/libp2p/specs/blob/98e0ca0e54fd83f7b8c86d7e9677e1e430da5810/addressing/README.md#multiaddr-in-libp2p
+    }
+
+    # Note: This is not quite valid IPLD Schema because the "structured" represenation does not exist yet, nor does the SingletonInt mechanism
+    # However, we can use this to describe that we have a union keyed by numbers rather than strings.
+    type TransferProtocol union {
+        | Bitswap
+        | GraphSync-FILv1

```

Comment:
Feature request after talking with @willscott it'd be nice if whichever union type we went with here we could add some fallback where servers could return data for transports they don't quite understand without needing someone to make a spec PR, update the implementations, etc. If clients don't know how to process the result, they can just throw it out.

In the structured case or keyed case it'd be nice to have an Any fallback for union cases we don't quite know (e.g. for strings we don't understand or for singletonInts we don't know about). We don't want to have to fallback to the less structured `type TransferProtocol struct { ID Bytes Hint optional Bytes # Or Any? }` but we can if we need to.

cc @petar @mvdan @warpfork

Reply:
https://github.com/ipld/ipld/issues/194

Reply:
It looks like there might be some implementation details in getting fallbacks merged into the schema schema, however there seems to be general alignment that this type of behavior is sane, so we'll use the syntax from the linked issue to represent the fallbacks required by the protocol.